File: libs/rust/observability/src/tracing.rs
#[derive(Clone, Debug, Eq, PartialEq)]
pub struct TraceContext { pub trace_id: String, pub span_id: String }
impl TraceContext { pub fn new(trace_id: &str, span_id: &str) -> Result<Self, &'static str> { if trace_id.len() != 32 || span_id.len() != 16 || !trace_id.chars().chain(span_id.chars()).all(|value| value.is_ascii_hexdigit() && !value.is_ascii_uppercase()) { return Err("trace identifiers must be lowercase hexadecimal"); } Ok(Self { trace_id: trace_id.into(), span_id: span_id.into() }) } }

File: libs/typescript/observability/src/tracing.ts
export interface TraceContext {
  readonly traceId: string;
  readonly spanId: string;
}
const trace = /^[0-9a-f]{32}$/;
const span = /^[0-9a-f]{16}$/;
export const traceContext = (traceId: string, spanId: string): TraceContext => {
  if (!trace.test(traceId) || !span.test(spanId))
    throw new Error("trace identifiers must be lowercase hexadecimal");
  return Object.freeze({ traceId, spanId });
};
export interface MetricValue {
  readonly name: string;
  readonly value: number;
}
export const metricValue = (name: string, value: number): MetricValue => {
  if (!/^[a-z][a-z0-9_.-]{0,127}$/.test(name) || !Number.isFinite(value))
    throw new Error("metric value is invalid");
  return Object.freeze({ name, value });
};
